### ISSUE TITLE
New Mockito stubbing api

### ISSUE THREAD

**Issue by mockitoguy** [mockito/mockito#303]
## Top Mockito priorities (equal importance):

I. Maximize test's clarity
II. Maximize test's failure output (easy test failure debugging)
III. Guide the engineer toward cleaner, higher quality tests that are easy to maintain
## Use cases, in the order of importance.

1) Unused stubbing is dead code, an irrelevant detail that clutters test's clarity.
2) Unintentionally unused stubbing can make test failures harder to debug.
3) When the stubbed call needs to be verified, it is declared twice in test (violation of DRY principle).
## Proposed solution

The solution consists of multiple changes. Perhaps it's better to show the code first, then describe the changes.

``` java
//JUnit rule automatically fails when stubs are unused:
MockitoJUnit.rule().strict();

//Offer a way to make specific stubbings optional:
given(mock.foo(), optional()).willReturn("bar");

//Offer a way to validate stubs for users who don't use JUnit rules:
Mockito.validateStubs();
```

**Summary:**

1) All stubbed method calls must be used (realized), otherwise the test fails.
2) Test failure is triggered by JUnit rule or Runner
  - When JUnit rule or runner is not used, new Mockito.verifyStubs() method can be used. The method is not intended for engineers to use in tests. It's intended for framework integrators (TestNG runners) or a common base class.
4) "optional" stubbings. Consider default stubbings that user may put in test constructor or setup method. Those default stubbings are used by most test methods but not all. This is a legit scenario that needs to be supported.
5) verifyNoMoreInteractions() no longer includes used stubs for validation **when** strict rule / runner is in use.

**Mockito 3:**

1. Strict rule by default
2. verifyNoMoreInteractions() no longer includes used stubs for validation

**Comment by ryan-gustafson** [comment]
It's unfortunate this isn't slated for 2.0 anymore, these are the kinds of features I'm most interested in seeing, as they help keep the tests accurate and complete.

In most cases, I'd like my mocks to:
- Not have magical default behavior, they should only do what I stub them to do, and nothing else. Failure to use a stubbing should fail fast, not lazily.  It would be great if the failure could indicate the "closest" stubbings to the actual call made, to help diagnose the misalignment.
- If I stubbed it, I usually intend it to be called, and therefore want it to be verified without repeating myself.

The lack of these kinds of features increases the time-cost of using Mockito because:
- After repeated refactorings by different people of varying abilities, the test stubbing can have almost _nothing_ at all to do with the actual code.  Time is wasted trying to understand what is really being tested, and reconciling the test setup against the actual code.
- Verifying interactions is highly repetitious to the point of being onerous.  It's really the only way to ensure the mock is being used in the expected fashion, so it should be far less effort.
- Default answers of null/zero/empty are fine if you care very little about the mock interactions, they are completely wrong when you do care.  If I only want the mock to have behaviors differing strictly by what was stubbed, I should get an exception at method call time, not a bogus response that the worst case could result in an indirect error very far from the call on the mock but _before_ you make it to the verify() call.  Debugging these kinds of problems can be a huge time sink, only to discover you missed a stubbing or it was slightly off.

**Comment by TimvdLippe** [comment]
This is fixed in #555

**Comment by szczepiq** [comment]
#555 only addresses some of the use cases. Plus, it only works when you use specific JUnit integration. The problems / use cases still stand and @ryan-gustafson provided _fantastic_ feedback.

We need more such feedback in order to make good decisions on how to improve Mockito API. @ryan-gustafson, we will need more info from you as we drill into those use cases :) We won't be able to include it in 2.0. However, the team has great momentum and plan to create 3.0 branch soon.

**Comment by szczepiq** [comment]
I updated the description, please give more feedback!

**Comment by robfletcher** [comment]
Might be a small point but `verifyStubbings` seems an odd name to me. Wouldn't the (admittedly blurred) distinction between a mock and a stub be that a latter _doesn't_ get verified but just supplies dummy return values? The current `verify` is clearer IMO.

**Comment by Sarev0k** [comment]
I think @ryan-gustafson brings up some great points, I too would prefer setting up and interacting with mocks that *never* return dummy values, and instead throw exceptions at method call time if the mock is interacted with in an unconfigured way.

In order to work around problems stemming from mocks returning default values, my team has decided to configure our mocks such that any arguments to methods of the right type will return test values, and then only after the code under test has executed, verify that the mocks were interacted with in the correct way.

Configuring our mocks such that only specific arguments return a test value has proved to be too much of a time sync for debugging issues caused by the default behavior of incorrectly configured mocks.

I think if implemented, @ryan-gustafson's suggestion would greatly simplify my team's test cases.

**Comment by TimvdLippe** [comment]
@Sarev0k did you check out the new Strict mode of the MockitoJunitRunner in Mockito 2? I think that would suit your usecase

**Comment by Sarev0k** [comment]
@TimvdLippe Unfortunately it does not.  Strict mode on MockitJunitRunner just reports unused mock methods.  It will not force the mock object to throw an exception if it is accessed in an unconfigured way (at least according to the documentation of this feature)

This means that if I do more tightly configure my mocks to exact parameters that I think should be passed, I would still have difficult to track down bugs, due to the default behavior of mock objects returning default values instead of throwing exceptions, when they are accessed in an unconfigured way.

Here's the interaction I'm looking for
```java
final Foo foo = mock(Foo.class);
given(foo.bar("test")).willReturn("response");
final String result = foo.bar("notconfigured"); // I would like this to throw, because bar("notconfigured") was never configured.
// The current behavior of Mockito is to return null above.  And unless I immediately dereference
// the result of this method, I now have a failing test with a NullPointerException in a completely
// different part of my code base.
```

**Comment by szczepiq** [comment]
@Sarev0k, great feedback. THANKS!

In what circumstances you want the unconfigured mock to throw:
a) always
b) only if the mock was stubbed
c) only if the mock was stubbed with specific method and that particular method is called with arguments that are different from the stubbing

If you want a) you can use create mocks with custom default Answer implementation that just fails. Do you guys us it? Is it convenient? Can we make it more convenient?

I respect this style of mocking and I want to make sure it is convenient to apply with Mockito. However, I am not convinced this is the best default behavior of Mockito. Strict mocking tends to lead to overspecified tests that get out of hand. The default lenient behavior may trigger more debugging but in turn tends to give cleaner tests. It's very hard to make a call what approach is better (and I'm sure you have an opinion on this :). We would need to hear a lot more feedback, from a lot more users to drastically change Mockito style.

**Comment by szczepiq** [comment]
@robfletcher "verifyStubbings" is horrible, I agree. Do you have other ideas? We could potentially fold it into existing "verifyNoMoreInteractions". The reason I suggest "verifyStubbings", despite awkward name, is that I don't see it as something used in a test, but rather in base class tear down method in case someone does not want to use rule or the runner.

This new API does not remove the existing "verify" API.

**Comment by Sarev0k** [comment]
@szczepiq I'm looking for it to always throw.  I know that Mockito supports a default answer that does this, however it has a few shortcomings.

1. Mockito doesn't provide a default answer implementation that throws.  Having a `Mockito.THROWS` constant as a peer to `Mockito.RETURNS_*` could be useful.
2. It's inconvenient to add a default answer to all your mock() invocations.  This litters your code, and someone is bound to forget to set their default answer correctly.
3. MockitoJunitRunner does not support custom default answers for the mocks that it creates.

I understand that making this the default behavior of Mockito is undesirable, however if there were some way to configure the default behavior of Mockito to use different behavior by default, I think that would be very desirable.

Perhaps a java System property that points to a class that contains the default answer that all mocked objects would use if provided.  Then I could just add `-Dmockito.mock.default.answer=com.my.default.answer` to my java invocation and all my mocks regardless of if they're created with the MockitoJunitRunner or explicit mock() invocations would all have the same default behavior.

**Comment by ryan-gustafson** [comment]
@szczepiq I've experimented with changing the default answer in Mockito 1.This interacts poorly with when/then style stubbing though.  Workarounds are to (a) use do/when style, which loses compile time  type safety on the returns, (b) temporarily disable the default answer behavior during setup, but this is a pain to manage and make sure you re-enable when done, or (c) wrap Mockito stubbing APIs to apply (b) automatically during the ongoing stubbing call.

This basically boils down to changing the answer behavior during stubbing to be different than a nonstubbing mock interaction.

I see Erik added some great comments whole I was writing this.

A separate point... 
I don't think standard Mockito behavior needs to change, as there are legitimate uses for wanting mocks with null defaults.  It would also be a huge shift for the user base.

That's said, I personally find the most arguments for using loose mocks in tests so they are not over specified to be logically unsound. If you don't know what was done with the mock, and you haven't mocked a behavior, then usually follows that the test is of the feel-good variety and not very good. Ultimately you can't really say what the code did other than it didn't blow up, produced a particular result in an unknown context, or had a few interactions on a mock. It's usually not possible to reason that there were no unexpected interactions that invalidate the test, or that the handwaving used when the test was first written still holds in the future. For these reasons I don't use Mockito like many apparently do, I want my mocks absolutely specified and fully verified whenever possible.  I blame my mathematics background for this desire for rigorousness.

When I don't or can't then I used Mockito like most people, but I make it clear in the test the intended bounds of this lack of rigorousness.  Such tests are usually rather shallow and only exist to provide loose constraints on the behavior of the code.  I usually think through such cases multiple times to ensure this is truly the proper strategy to use, as the test is inherently ill-defined and potentially invalid now or in the future, even it if it keeps on passing the entire time.

**Comment by szczepiq** [comment]
Thank you for all feedback!

I think we have deviated from the core use cases so I forked #649.

Guys, what about void methods? Do you also want to pre-configure them? If that's the case what you are really asking is EasyMock / JMock style of mocking where every interaction is configured upfront. Let's continue the discussion in #649.

**Comment by szczepiq** [comment]
For configuring Mockito 'globally' per code base, you can use: http://site.mockito.org/mockito/docs/current/org/mockito/configuration/IMockitoConfiguration.html

**Comment by szczepiq** [comment]
I brainstormed the use cases, here’s the idea. Next time I allocate time to this ticket, I will update it for clarity. Feedback is very welcome.

Target plan (not exhaustive), breaking changes annotated with (*):

1. (*) Unused stubbings fail the test (via runner or rule, additionally there is an explicit method provided when someone does not use rule/runner)
2. Ability to make stubbings "optional". Example new api: optional() prefix: 
```java
optional().when(mock.foo()).thenReturn(x)
```
3. New Answers.Strict for creating strict mocks - fail when unstubbed non-void method invoked
4. (*) When method ‘foo’ is stubbed with arguments, then called with different argument in code fail immediately, unless optional()
5. Some way to temporarily enable strict mocks for debugging a single test method (need more design and fleshing out)

Example stepping stones (compatible with 2.x so can be worked on):

1. Add Strict mock - #649
2. Add MockitoHint when stub used with different arguments
     - some way to suppress hint
     - hint even if JUnit rule/runner is not used
3. Strict JUnit rule option
     - unused stubbings fail at the end of test
     - stubbed method invoked with mismatching args fails immediately unless optional()
4. Create new optional() prefix

**Comment by szczepiq** [comment]
I'm working on coming up with a cohesive plan for "strictness" in Mockito. Create new ticket at #769.

**Comment by TimvdLippe** [comment]
Closing per the above comment.
